thetis/utility.py

```diff
@@ -1004,3 +1007,162 @@ def solve(self, solve_w=True):
+        # Ensure that key_order contains the correct set of keys
+        expected_keys = set(model_dict['species'].keys())
+        assert len(key_order) == len(expected_keys)
+        assert set(key_order) == expected_keys
+
+    return parse_tracers_from_dict(
+        model_dict, function_space, lambdify_modules=lambdify_modules,
+        key_order=key_order, append_dimension=append_dimension)
+
+
+def parse_tracers_from_dict(model_dict, function_space, lambdify_modules=None,
+                            key_order=None, append_dimension=False):
+    r"""
+    Constructs and returns an ordered dictionary that specifies an ADR model.
+
+    :arg model_dict: Dictionary defining species in an ADR system.
+        This must include the keys used to identify each species, their
+        full names, their diffusivities and their reaction terms.
+        Any hard-coded numeric values should be defined under
+        model_dict['constants'], and the details of each species should be
+        defined under model_dict['species'].
+        A dictionary obtained from a correctly-formatted YAML file using
+        `yaml.safe_load` should be sufficient.
+
+        **Example**
+
+        .. code-block:: python
+
+            {
+                'constants': {
+                    'D1': 8e-05,
+                    'D2': 4e-05,
+                    'k1': 0.024,
+                    'k2': 0.06
+                },
+                'species': {
+                    'a': {
+                        'diffusion': 'D1',
+                        'name': 'Tracer A',
+                        'reaction': '-a*b**2 + k1*(1-a)'
+                    },
+                    'b': {
+                        'diffusion': 'D2',
+                        'name': 'Tracer B',
+                        'reaction': 'a*b^2 - (k1+k2)*b'
+                    }
+                }
+            }
+
+    :arg function_space: The Firedrake :class:`FunctionSpace` in which the
+        :class:`Function` for each tracer will reside.
+    :kwarg lambdify_modules: A string or dictionary to be passed as the
+        `modules` parameter of `sympy.lambdify`.
+    :kwarg key_order: List of species keys. If not `None`, this
+        determines the order in which species are added to the
+        :class:`OrderedDict` before it is returned.
+    :kwarg append_dimension: If `True`, a suffix will be appended to the
+        label of each species from `model_dict`. The form of the suffix is
+        `_Nd`, where `N` is the dimensionality of the spatial domain in
+        which `function_space` resides.
+    """
+    key_suffix = '_%id' % function_space.mesh().topological_dimension() \
+                 if append_dimension else ''
+    mdl = ADR_Model(model_dict, lambdify_modules=lambdify_modules)
+
+    species = key_order
+    if species is None:
+        # Get tracer labels
+        species = mdl.list_species_keys()
+
+    adr_model = OrderedDict({s + key_suffix: {} for s in species})
```

Comment:
Another point for discussion: should the `ADR_Model` instance be used internally to set up an `OrderedDict` as it is now, or should the `ADR_Model` class be updated to allow it to be used instead of an `OrderedDict`? (e.g. by passing a `FunctionSpace` to the constructor and having it set up `Functions`, `Constant`s and UFL expressions internally)

Reply:
I think it would be neater if `ADR_Model` had an `OrderedDict`-like structure itself. Would that be feasible?